Add farmbot case (#1431)
elixir-lang/elixir-lang.github.com: _posts/2020-08-20-embedded-elixir-at-farmbot.markdown

---
layout: post

> title: Embedded Elixit at Farmbot

Reviewer: ```suggestion
title: Embedded Elixir at Farmbot
```

Reviewer: this one is still missing

author: José Valim
category: Cases
excerpt: A case study of how Elixir is being used at Farmbot.
---

> [FarmBot](https://farm.bot/) is an open-source precision agriculture CNC farming project that includes a robot farming machine, software, and documentation including a farming data repository. FarmBot’s machines use IoT technology to make it easy for farmers to remotely manage their gardens.

Reviewer: ```suggestion
[FarmBot](https://farm.bot/) is an open-source precision agriculture CNC farming project that includes a robot farming machine, software, and documentation including a farming data repository. FarmBot's machines use IoT technology to make it easy for farmers to remotely manage their gardens.
```
Replace `’` with single quote. Here and everywhere throughout the text. I am just making the suggestion in this occurrence.

> Farmbot is built with [Nerves](https://www.nerves-project.org/), an open-source platform and infrastructure you need to build, deploy, and securely manage your fleet of IoT devices at speed and scale.

Author: ```suggestion
Farmbot is built with [Nerves](https://www.nerves-project.org/), an open-source platform and infrastructure to build, deploy, and securely manage your fleet of IoT devices at speed and scale.
```

![Farmbot](/images/cases/farmbot.jpg)

## How Nerves helped

Connor brought Nerves to Farmbot by porting the entire proof-of-concept Ruby application he'd created over to Nerves, which he did in his free time over the course of a month, taking him about 20 hours total. He also continued to make open-source contributions to Nerves, helping to structure the networking functionality that is now part of [NervesHub](https://www.nerves-hub.org/), the extensible web service that enables over-the-air firmware update management.

<blockquote style="font-size: 24px; color: #444">
<p>The biggest benefit of using Nerves is definitely how fast you can get up and running.</p>
<p>— Connor Rigby, Embedded Systems Engineer</p>
</blockquote>

Connor says that the Nerves Platform and what eventually became NervesHub was a great choice for Farmbot because:

### 1. Nerves supports lean systems and operates well in low-bandwidth areas

> Because Nerves bundles entire applications into relatively small archives in terms of firmware images for full Linux systems, Farmbot can use NervesHub to send over-the-air updates more quickly and users can download them faster. (For comparison, an Android update generally clocks in at around 4 GB, but a Nerves update can be packed into as little as 12 MB.) 

Reviewer: ```suggestion
Because Nerves bundles entire applications into relatively small archives in terms of firmware images for full Linux systems, Farmbot can use NervesHub to send over-the-air updates more quickly and users can download them faster. For comparison, an Android update generally clocks in at around 4 GB, but a Nerves update can be packed into as little as 12 MB. 
```

> This is especially helpful for Farmbot users who operate in more remote locations with lower bandwidth and less reliable access to WiFi. When an internet connection is available, NervesHub will connect and check if there's an update, then prompt the user to install the update. 

Reviewer: ```suggestion
This is especially helpful for Farmbot users who operate in more remote locations with lower bandwidth and less reliable access to Wi-Fi. When an internet connection is available, NervesHub will connect and check if there's an update, and then prompt the user to install the update. 
```

Reviewer: Alternatively it could be ", to then prompt"

> This is especially helpful for Farmbot users who operate in more remote locations with lower bandwidth and less reliable access to WiFi. When an internet connection is available, NervesHub will connect and check to see if there's an update, then prompt the user to install the update — all without Connor needing to put in extra time to make it happen, because it's already set up in NervesHub. 

Reviewer: Connor, or the user?

Author: ```suggestion
This is especially helpful for Farmbot users who operate in more remote locations with lower bandwidth and less reliable access to WiFi. When an internet connection is available, NervesHub will connect and check to see if there's an update, then prompt the user to install the update. 
```

> This is especially helpful for Farmbot users who operate in more remote locations with lower bandwidth and less reliable access to WiFi. When an internet connection is available, NervesHub will connect and check to see if there's an update, then prompt the user to install the update — all without Connor needing to put in extra time to make it happen, because it's already set up in NervesHub. 

Reviewer: ```suggestion
This is especially helpful for Farmbot users who operate in more remote locations with lower bandwidth and less reliable access to WiFi. When an internet connection is available, NervesHub will connect and check if there's an update, then prompt the user to install the update — all without Connor needing to put in extra time to make it happen, because it's already set up in NervesHub. 
```

While correct, I find it a bit redundant that expression.

> ### 2. Nerves adds convenience with low overhead.

Reviewer: ```suggestion
### 2. Nerves adds convenience with low overhead
```

> For devices that are already connected to the internet, connecting to Nerves requires no additional configuration because NervesHub is compatible with the current public key infrastructure for device-to-cloud communication. Since Farmbot already had internet-connected devices when they brought Nerves onboard, they were able to use the same “key” to sign in to NervesHub that they use for their cloud service. 

Reviewer: ```suggestion
For devices that are already connected to the internet, connecting to Nerves requires no additional configuration because NervesHub is compatible with the current public key infrastructure for device-to-cloud communication. Since Farmbot already had internet-connected devices when they brought Nerves onboard, they were able to use the same "key" to sign in to NervesHub that they use for their cloud service. 
```
Same here, use regular double-quotes

> ### 3. Nerves has all the benefits of Elixir and Erlang.

Reviewer: ```suggestion
### 3. Nerves has all the benefits of Elixir and Erlang
```
